modules/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/BLangVM.java

```diff
@@ -1489,8 +1494,22 @@ private void execStoreOpcodes(StackFrame sf, int opcode, int[] operands) {
                     handleNullRefError();
                     break;
                 }
-
-                bMap.put(sf.stringRegs[j], sf.refRegs[k]);
+                BMapType mapType = (BMapType) bMap.getType();
+                if (sf.refRegs[k] == null) {
+                    if (!isValueType(mapType.getConstrainedType())) {
+                        // TODO improve logic for null case
+                        bMap.put(sf.stringRegs[j], sf.refRegs[k]);
+                    } else {
+                        throw BLangExceptionHelper.getRuntimeException(RuntimeErrors.INVALID_MAP_INSERTION,
```

Comment:
This is not the correct way of handling the error. You may define an error called InvalidMapInsertionException in ballerina.built and handle it similar to "handleNullRefError".